ddtrace/monkey.py

```diff
@@ -202,7 +210,7 @@ def _patch_module(module):
         except AttributeError:
             # if patch() is not available in the module, it means
             # that the library is not installed in the environment
-            raise PatchException("module '%s' not installed" % module)
+            raise ModuleNotFoundException("module '%s' not installed" % module)
```

Comment:
Is it safe to assume that any `AttributeError` error coming from `_patch_module` will mean the module couldn't be loaded? It might be more dependable/trustworthy if the patch modules themselves raised a `ModuleNotFoundException` in the portion of their code that actually does the module lookup e.g. on `required_modules` returning one or more missing modules.

Reply:
Thanks for taking a look and good question! It's technically not for the reasons you pointed out. Test coverage most likely eliminates the possibility that an AttributeError could be raised on `patch()`, but we shouldn't depend on our tests covering every scenario. Looking at this code again, it's also silly that both the import and patch statements are in the same `try` so definitely some more work to be done here 🙂. I like the idea of `required_modules` raising.

Reply:
Yep, I agree, `imported_module.patch()` would've been better in its own try. But there's still the problem that `AttributeError` could be thrown from inside `imported_module.patch()` (which I poorly stated in my original comment).  In that scenario, if we raised a `ModuleNotFoundException`, it would be misleading because the module to be patched was indeed found but some other AttributeError` error occurred while patching. Hence the idea of having the actual module look up logic raise `ModuleNotFoundException`.

Reply:
Oh yeah no, I totally understood from your first comment 😄. I took a look at `require_modules` but unfortunately our usage of it doesn't make sense to have it raise, although agreed that it would be so much more ideal.

I've now separated out the statements and now just do an actual `hasattr` check on the module for the patch attribute rather than depending on catching an AttributeError on the look up.

Reply:
Sorry for over explaining myself lol. Checking for the `patch` attribute explicitly is a great solve for the case where `AttributeError` is thrown from within the `patch` method. And thanks for your help; I'm looking forward to not seeing that exception in our logs anymore!

Hopefully in the future these patching modules will have a clearer way to express that they aren't able to do the patching other than just not having the patch attribute. It's probably just a personal preference but it feels like the `patch` method should always exist but return a result or something that indicates the status of the patch.

Reply:
Haha no worries 🙂 ! I'm in complete agreement with you, it's quite an odd behaviour which I'm not a fan of either (I don't think exported objects should be calculated at run-time).

To give you some insight to our plans, we want to transition to not importing integration libraries at all within the library and instead do a patch-on-import solution which will install the instrumentation only when the library would normally be imported in a user's application.